File: 375.guessNumberHigherOrLowerII/guessNumberHigherOrLowerII.java
class Solution {
    public int getMoneyAmount(int n) {
        int[][] dp = new int[n + 1][n + 1];
        return dfs(dp, 1, n);
    }
    
    public int dfs(int[][] dp, int start, int end) {
        if (start == end) return dp[start][end] = 0;
        if (dp[start][end] != 0) return dp[start][end];
        int cur = Integer.MAX_VALUE;
        for (int i = start; i <= end; i++) {
            if (i == start) {
                cur = Math.min(cur, i + dfs(dp, i + 1, end));
            } else if (i == end) {
                cur = Math.min(cur, i + dfs(dp, start, i - 1));
            } else {
                cur = Math.min(cur, i + Math.max(dfs(dp, start, i - 1), dfs(dp, i + 1, end)));
            }
        }
        return dp[start][end] = cur;
    }
}

File: 375.guessNumberHigherOrLowerII/guessNumberHigherOrLowerII.cpp
// clang++ guessNumberHigherOrLowerII.cpp -std=c++11

#include <iostream>
#include <vector>
using namespace std;


class Solution {
public:
    int getMoneyAmount(int n) {
        vector<vector<int> > dp(n + 1, vector<int>(n + 1, -1));
        return helper(1, n, dp);
    }
    int helper(int start, int end, vector<vector<int> >& dp) {
        if (start >= end)
            return 0;
        if (dp[start][end] != -1)
            return dp[start][end];
        int res = INT_MAX;
        for (int i = start; i <= end; i++) {
            res = min (res, max(helper(start, i - 1, dp), helper(i + 1, end, dp)) + i);
        }
        return dp[start][end] = res;
    }
};

int main()
{
    Solution s;
    vector<vector<int> > pairs = {{1, 2},
                                  {2, 3},
                                  {3, 4}};
    cout << s.getMoneyAmount(10) << endl;
    return 0;
}
